pennylane_lightning_gpu/src/util/MPIManager.hpp

```diff
@@ -0,0 +1,612 @@
+    MPI_Abort(MPI_COMM_WORLD, 1);
+}
+
+#define PL_MPI_IS_SUCCESS(fn)                                                  \
+    {                                                                          \
+        int errcode;                                                           \
+        errcode = (fn);                                                        \
+        if (errcode != MPI_SUCCESS)                                            \
+            errhandler(errcode, #fn);                                          \
+    }
+
+template <typename T> auto cppTypeToString() -> const std::string {
+    const std::string typestr = std::type_index(typeid(T)).name();
+    return typestr;
+}
+
+/**
+ * @brief MPI operation class. Maintains MPI related operations.
+ */
+class MPIManager {
+  private:
+    bool isExternalComm_;
+    size_t rank_;
+    size_t size_per_node_;
+    size_t size_;
+    MPI_Comm communicator_;
+
+    std::string vendor_;
+    size_t version_;
+    size_t subversion_;
+
+  public:
+    MPIManager() : communicator_(MPI_COMM_WORLD) {
+        isExternalComm_ = true;
+        int rank_int;
+        int size_int;
+        PL_MPI_IS_SUCCESS(MPI_Comm_rank(communicator_, &rank_int));
+        PL_MPI_IS_SUCCESS(MPI_Comm_size(communicator_, &size_int));
+
+        rank_ = static_cast<size_t>(rank_int);
+        size_ = static_cast<size_t>(size_int);
+
+        findVendor();
+        findVersion();
+        getNumProcsPerNode();
+        check_mpi_config();
+    }
+
+    MPIManager(MPI_Comm communicator) : communicator_(communicator) {
+        isExternalComm_ = true;
+        int rank_int;
+        int size_int;
+        PL_MPI_IS_SUCCESS(MPI_Comm_rank(communicator_, &rank_int));
+        PL_MPI_IS_SUCCESS(MPI_Comm_size(communicator_, &size_int));
+
+        rank_ = static_cast<size_t>(rank_int);
+        size_ = static_cast<size_t>(size_int);
+
+        findVendor();
+        findVersion();
+        getNumProcsPerNode();
+        check_mpi_config();
+    }
+
+    MPIManager(int argc, char **argv) {
+        PL_MPI_IS_SUCCESS(MPI_Init(&argc, &argv));
+        isExternalComm_ = false;
+        communicator_ = MPI_COMM_WORLD;
+        int rank_int;
+        int size_int;
+        PL_MPI_IS_SUCCESS(MPI_Comm_rank(communicator_, &rank_int));
+        PL_MPI_IS_SUCCESS(MPI_Comm_size(communicator_, &size_int));
+
+        rank_ = static_cast<size_t>(rank_int);
+        size_ = static_cast<size_t>(size_int);
+
+        findVendor();
+        findVersion();
+        getNumProcsPerNode();
+        check_mpi_config();
+    }
+
+    MPIManager(const MPIManager &other) {
+        isExternalComm_ = true;
+        rank_ = other.rank_;
+        size_ = other.size_;
+        communicator_ = other.communicator_;
+        vendor_ = other.vendor_;
+        version_ = other.version_;
+        subversion_ = other.subversion_;
+        size_per_node_ = other.size_per_node_;
+    }
+
+    virtual ~MPIManager() final {
```

Comment:
Are we certain this is to be final? If so, I'd suggest adding it to the class definition above also.

Reply:
As discussed previously, the MPIManager class should be final. Added final to the class definition.